espnet2/speechlm/loss.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
+
+# Copyright 2025 Jinchuan Tian
+# Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+import logging
+
+import torch
+
+
+class SpeechLMCrossEntropyLoss(torch.nn.Module):
+    def __init__(
+        self,
+        pad,
+        token_bias,
+        modality_weights,
+        image_interval_split,
+        loss_type="mean",
+        lm_head: torch.nn.Linear = None,
+    ):
+        super().__init__()
+
+        self.pad = pad
+        self.lm_head = lm_head
+        self.loss_type = loss_type
+
+        # (1) loss weight
+        vocab_size = lm_head.weight.size(0)
+        self.weight = torch.ones(vocab_size).float()
+        self.weight.requires_grad_(False)
+        for name, m_weight in modality_weights.items():
+            if name not in token_bias:
+                logging.warning(f"Modality {name} not in token_bias. Skip it")
+                continue
+
+            start, end = token_bias[name]
+            self.weight[start:end] = m_weight
+
+        # (2) aux loss interval
+        self.aux_loss_interval = []
+        # Codec codebook is small, put them in one interval for efficiency
+        if "codec" in token_bias:
+            self.aux_loss_interval.append(token_bias["codec"])
+        # Image codebook is large, compute them separately
+        if "image" in token_bias:
+            start, end = token_bias["image"]
+            assert (end - start) % image_interval_split == 0
+            inc = (end - start) // image_interval_split
+            i = 0
+            while start + i * inc < end:
+                self.aux_loss_interval.append(
+                    (
+                        start + i * inc,
+                        start + (i + 1) * inc,
+                    )
+                )
+                i += 1
+
+    def forward(self, hidden, targets, loss_mask):
```

Comment:
also a unit test for this, since its susceptible to breaking as people add more modalities. I think we just need to check if the intervals and masking are correct for each loss